[Screen Time Refactoring] macOS support for URL donations

jesxilin opened the pull request:
#### 9b4b2aab8589840de71c3f65035c20b10057ce66
<pre>
[Screen Time Refactoring] macOS support for URL donations
<a href="https://bugs.webkit.org/show_bug.cgi?id=283944">https://bugs.webkit.org/show_bug.cgi?id=283944</a>
<a href="https://rdar.apple.com/140438800">rdar://140438800</a>

Reviewed by Aditya Keerthi and Tim Horton.

Add a flag/preference to enable Screen Time donation.
Implemented Screen Time donation for loaded pages.
Installed Screen Time shield view to block web content when needed.
Halts media playback when web content is blocked
and shield is shown (not supported for PIP yet).

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PlatformMac.cmake:
* Source/WebCore/PAL/pal/cocoa/ScreenTimeSoftLink.h: Added.
* Source/WebCore/PAL/pal/cocoa/ScreenTimeSoftLink.mm: Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _installScreenTimeWebpageController]):
(-[WKWebView _uninstallScreenTimeWebpageController]):
(-[WKWebView observeValueForKeyPath:ofObject:change:context:]):
(-[WKWebView dealloc]):
(-[WKWebView _screenTimeWebpageController]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::installScreenTimeWebpageController):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::prepareToLoadWebPage):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::installScreenTimeWebpageController):
(WebKit::updateScreenTimeWebpageControllerURL):
(WebKit::PageClientImpl::didCommitLoadForMainFrame):

Canonical link: <a href="https://commits.webkit.org/287638@main">https://commits.webkit.org/287638@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/176d9c4f419122d2e4a60256dc9fc252a0eaaaa9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80101 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59103 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33513 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84618 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31077 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82212 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68164 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7401 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62629 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20450 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83170 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52701 "Found 1 new test failure: media/audioSession/getUserMedia.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72958 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42933 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50030 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27112 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29537 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/73077 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71158 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27619 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86050 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/79158 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7321 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5187 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70900 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7496 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68797 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70135 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17512 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14135 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13082 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/101566 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7285 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12802 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24758 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7124 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10644 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8929 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -226,6 +226,8 @@
 static const BOOL defaultFastClickingEnabled = NO;
 #endif
 
+static void *readyForDisplayKVOContext = &readyForDisplayKVOContext;
```
This name seems wrong? "Ready for display" is an AVKit KVO property; you should rename this after your own!

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -338,6 +340,56 @@ static uint32_t convertSystemLayoutDirection(NSUserInterfaceLayoutDirection dire
+}
+
+#if ENABLE(SCREEN_TIME)
+- (void)_deallocScreenTimeWebpageController
```
"Dealloc" means something specific, which this doesn't actually do. Maybe "uninstall"?

hortont424 replied:
And if you do that maybe the other one becomes "install"?

jesxilin replied:
Will change!

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -28,6 +28,7 @@
 #ifdef __cplusplus
 
 #import "PDFPluginIdentifier.h"
+#import <ScreenTime/STWebpageController.h>
```
Sorting

jesxilin replied:
I got a warning from the checker, smth about alphabetical ordering. Is it still in the wrong place?

whsieh replied:
I'd expect this to be lower down, below `#import <wtf/spi/cocoa/NSObjCRuntimeSPI.h>`, in its own `#if`-section

```
#if ENABLE(SCREENTIME)
#import <ScreenTime/STWebpageController.h>
#endif
```

jesxilin replied:
Why is that so? I thought it would be under `#import "PDFPluginIdentifier.h"` and before the Ws b/c of the alphabetical ordering.

rr-codes replied:
it should be guarded, and guarded imports are typically separated from the rest

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -262,6 +263,7 @@ struct PerWebProcessState {
     NSUInteger _partialIntelligenceTextAnimationCount;
     BOOL _writingToolsTextReplacementsFinished;
 #endif
+    RetainPtr<STWebpageController> _screenTimeWebpageController;
```
Guarding?

whsieh replied:
I think you could also consider just keeping track of a single `RetainPtr<STWebpageController>` ivar/member variable, which means deleting either this ivar (and using `_page->screenTimeWebpageController()`) or deleting `m_screenTimeWebpageController` on `WebPageProxy` (and instead adding a `PageClient` hook that asks the web view for the controller).

That way it's easier to avoid any state where the `WKWebView`'s ST controller is out of sync with the `WebPageProxy`'s controller.

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2015,6 +2086,8 @@ - (WKNavigation *)loadFileRequest:(NSURLRequest *)request allowingReadAccessToUR
     isAppInitiated = request.attribution == NSURLRequestAttributionDeveloper;
 #endif
 
+    [self _initializeScreenTimeWebpageController];
```
Does this really need to be sprinkled in every load method?

whsieh replied:
It _might_ be cleaner to add a method to `PageClient`, that's called from inside `WebPageProxy::prepareToLoadWebPage`. The page client method could then call  `-[WKWebView _initializeScreenTimeWebpageController]` — that way, all ways of navigating web views would be covered.

hortont424 commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -6311,6 +6311,20 @@ ScreenOrientationLockingAPIEnabled:
+  status: unstable
+  condition: ENABLE(SCREEN_TIME)
+  humanReadableName: "Screen Time"
+  humanReadableDescription: "Enable Web ScreenTime"
```
I'm not sure what the word Web is buying us here, given this is WebKit. Also, let's pick one spelling/spacing of Screen Time (ideally the one used in e.g. Settings)

hortont424 commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -266,7 +266,7 @@ WK_SPRINGBOARDSERVICES_LDFLAGS = $(WK_SPRINGBOARDSERVICES_LDFLAGS_$(WK_PLATFORM_
 WK_SPRINGBOARDSERVICES_LDFLAGS_iphonesimulator = -framework SpringBoardServices;
 
-FRAMEWORK_AND_LIBRARY_LDFLAGS = -lobjc -framework CFNetwork -framework CoreAudio -framework CoreFoundation -framework CoreGraphics -framework CoreText -framework Foundation -framework ImageIO -framework IOKit -framework IOSurface -framework WebKitLegacy -lnetwork -framework Metal $(WK_APPSERVERSUPPORT_LDFLAGS) $(WK_ACCESSIBILITY_LDFLAGS) $(WK_APPKIT_LDFLAGS) $(WK_APPLE_PUSH_SERVICE_LDFLAGS) $(WK_ ...
+FRAMEWORK_AND_LIBRARY_LDFLAGS = -lobjc -framework CFNetwork -framework CoreAudio -framework CoreFoundation -framework CoreGraphics -framework CoreText -framework Foundation -framework ImageIO -framework IOKit -framework IOSurface -framework WebKitLegacy -framework ScreenTime -lnetwork -framework Metal $(WK_APPSERVERSUPPORT_LDFLAGS) $(WK_ACCESSIBILITY_LDFLAGS) $(WK_APPKIT_LDFLAGS) $(WK_APPLE_PUSH_ ...
```
I expect ScreenTime.framework is not available in all configurations, we should make sure this is actually workable, and use some other kind of linking if not

whsieh replied:
Yeah, you might need to introduce a `WK_SCREENTIME_LDFLAGS` similar to the other build settings above, which would only link ScreenTime on platforms where it's available (and/or required).

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -722,6 +778,11 @@ - (void)addObserver:(NSObject *)observer forKeyPath:(NSString *)keyPath options:
 
+- (void)removeObserver:(NSObject *)observer forKeyPath:(NSString *)keyPath context:(void *)context
+{
+    [super removeObserver:observer forKeyPath:keyPath context:context];
```
What is this method for? It seems to just call super?

jesxilin replied:
Yes, I am removing it now though.

hortont424 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -1882,6 +1883,9 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     void handleControlledElementIDResponse(const String&) const;
     bool isPlayingVideoInEnhancedFullscreen() const;
 
+    RetainPtr<STWebpageController> screenTimeWebpageController() const { return m_screenTimeWebpageController; };
```
Lots of guarding

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -338,6 +340,56 @@ static uint32_t convertSystemLayoutDirection(NSUserInterfaceLayoutDirection dire
+        [_screenTimeWebpageController addObserver:self forKeyPath:@"URLIsBlocked" options:(NSKeyValueObservingOptionOld | NSKeyValueObservingOptionNew) context:&readyForDisplayKVOContext];
+#if PLATFORM(MAC)
+        NSView *screenTimeView = [_screenTimeWebpageController view];
+        screenTimeView.frame = self.bounds;
```
Who keeps the frame up to date if the web view resizes?

whsieh replied:
We could add some logic in here:

```
void WebViewImpl::setFrameSize(CGSize)
{
    [m_layoutStrategy didChangeFrameSize];

    auto updatedBounds = [m_view bounds];
    [m_warningView setFrame:updatedBounds];
    [m_page->screenTimeWebpageController() view].frame = updatedBounds;
}
```

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -1882,6 +1883,9 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     bool isPlayingVideoInEnhancedFullscreen() const;
 
+    RetainPtr<STWebpageController> screenTimeWebpageController() const { return m_screenTimeWebpageController; };
+    void setScreenTimeWebpageController(RetainPtr<STWebpageController> st) { m_screenTimeWebpageController = st; }
```
Minor nit - we generally avoid abbreviations like `st`, even in cases like this.
```suggestion
    void setScreenTimeWebpageController(RetainPtr<STWebpageController>&& controller) { m_screenTimeWebpageController = WTFMove(controller); }
```

whsieh commented on Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
```diff
@@ -293,6 +293,19 @@ - (BOOL)_hostsLayersInWindowServer;
+#if ENABLE(SCREEN_TIME)
+    auto webView = this->webView();
+    NakedPtr<WebKit::WebPageProxy> pageProxy = [webView _page];
+    if (RetainPtr<STWebpageController> screenTimeWebpageController = pageProxy->screenTimeWebpageController()) {
```
```suggestion
    NakedPtr pageProxy = [webView _page];
    if (RetainPtr screenTimeWebpageController = pageProxy->screenTimeWebpageController()) {
```

Nit - I think you can omit the template types here.

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -338,6 +340,56 @@ static uint32_t convertSystemLayoutDirection(NSUserInterfaceLayoutDirection dire
+        else if ([urlWasBlocked boolValue] == YES && [urlIsBlocked boolValue] == NO)
+            [self setAllMediaPlaybackSuspended:NO completionHandler:nil];
+    } else
+        [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
```
I think we could simplify this logic a bit:

```suggestion
    if (context == &screenTimeControllerBlockedKVOContext) {
        BOOL wasBlocked = dynamic_objc_cast<NSNumber>(change[NSKeyValueChangeOldKey]).boolValue;
        BOOL isBlocked = dynamic_objc_cast<NSNumber>(change[NSKeyValueChangeNewKey]).boolValue;
        if (wasBlocked != isBlocked)
            [self setAllMediaPlaybackSuspended:isBlocked completionHandler:nil];
        return;
    }
    [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
```

hortont424 commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -266,7 +266,13 @@ WK_SPRINGBOARDSERVICES_LDFLAGS = $(WK_SPRINGBOARDSERVICES_LDFLAGS_$(WK_PLATFORM_
+WK_SCREENTIME_LDFLAGS = $(WK_SCREENTIME_LDFLAGS_$(WK_PLATFORM_NAME));
+WK_SCREENTIME_LDFLAGS_iphoneos = -framework ScreenTime;
+WK_SCREENTIME_LDFLAGS_iphonesimulator = -framework ScreenTime;
+WK_SCREENTIME_LDFLAGS_macosx = -framework ScreenTime;
```
I still wonder if this is actually available in all kinds of macOS and iOS builds, or if this is going to break some obscure build. Richard (or any of us, but especially Richard) can help with commands to confirm :)

rr-codes replied:
indeed. I'll write up a paragraph or so explaining all of this in a bit.

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -338,6 +340,58 @@ static uint32_t convertSystemLayoutDirection(NSUserInterfaceLayoutDirection dire
+        NSView *screenTimeView = [_screenTimeWebpageController view];
+        screenTimeView.frame = self.bounds;
+        screenTimeView.translatesAutoresizingMaskIntoConstraints = NO;
+        // add to webview
```
These are "what" comments. They're also not complete sentences. In WebKit, we usually A) prefer code that is so obvious it doesn't need comments, and B) if it really does, they should be "why" comments, in complete sentences, with capitalization and punctuation.

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -338,6 +340,58 @@ static uint32_t convertSystemLayoutDirection(NSUserInterfaceLayoutDirection dire
+#endif // PLATFORM(MAC)
+    }
+}
+#endif // ENABLE(SCREEN_TIME)
```
why end the block and then reopen with the same condition?

jesxilin replied:
Will change!

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -795,6 +853,7 @@ - (WKNavigation *)loadRequest:(NSURLRequest *)request
     THROW_IF_SUSPENDED;
     if (_page->isServiceWorkerPage())
         [NSException raise:NSInternalInconsistencyException format:@"The WKWebView was used to load a service worker"];
+
```
I think you can revert this change

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -1973,6 +2037,7 @@ - (void)_didChangeEditorState
 - (WKNavigation *)loadSimulatedRequest:(NSURLRequest *)request response:(NSURLResponse *)response responseData:(NSData *)data
 {
     THROW_IF_SUSPENDED;
+
```
here too

hortont424 commented on Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
```diff
@@ -293,6 +293,19 @@ - (BOOL)_hostsLayersInWindowServer;
+    auto webView = this->webView();
+    NakedPtr<WebKit::WebPageProxy> pageProxy = [webView _page];
+    if (RetainPtr screenTimeWebpageController = [webView _screenTimeWebpageController]) {
+        // If there is a preexisting controller but Screen time is disabled, remove controller.
```
Screen Time (capitalization?)

this is to make dynamically disabling the setting work for an existing view? I don't think it's strictly necessary that that works, but I guess it's nice that you considered it!

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -338,6 +340,58 @@ static uint32_t convertSystemLayoutDirection(NSUserInterfaceLayoutDirection dire
+
+- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey, id> *)change context:(void *)context
+{
+    if (context == &screenTimeWebpageControllerBlockedKVOContext) {
```
I think this if statement and its contents should also be guarded, along with the screenTimeWebpageControllerBlockedKVOContext variable itself

jesxilin replied:
Will do!

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -673,6 +727,10 @@ - (void)dealloc
 
+#if ENABLE(SCREEN_TIME)
+    [self _uninstallScreenTimeWebpageController];
+#endif // ENABLE(SCREEN_TIME)
```
no need for the `// ENABLE(SCREEN_TIME)` since it's a trivial guarding

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -1862,6 +1921,11 @@ - (WKPageRef)_pageForTesting
     return _page.get();
 }
 
+- (RetainPtr<STWebpageController>)_screenTimeWebpageController
```
I think typically we just return the raw reference like `STWebpageController *` and then the implementation will be `return _screenTimeWebpageController.get()`

rr-codes commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -1882,6 +1883,7 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     void handleControlledElementIDResponse(const String&) const;
     bool isPlayingVideoInEnhancedFullscreen() const;
 
+
```
undo

rr-codes commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2383,6 +2385,7 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     void didDestroyNotification(const WTF::UUID& notificationID);
     void pageWillLikelyUseNotifications();
 
+
```
undo

rr-codes commented on Source/WebKit/UIProcess/mac/PageClientImplMac.h:
```diff
@@ -325,6 +325,10 @@ class PageClientImpl final : public PageClientImplCocoa
     void requestScrollToRect(const WebCore::FloatRect& targetRect, const WebCore::FloatPoint& origin) override;
 
+#if ENABLE(SCREEN_TIME)
+        void installScreenTimeWebpageController() override;
```
I thinkkk there's extra indentation here? not sure

rr-codes commented on Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
```diff
@@ -1147,6 +1160,13 @@ static inline NSCorrectionResponse toCorrectionResponse(AutocorrectionResponse r
+#if ENABLE(SCREEN_TIME)
+void PageClientImpl::installScreenTimeWebpageController()
+{
+    [this->webView() _installScreenTimeWebpageController];
```
I think we generally exclude `this->`

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -472,6 +485,9 @@ struct PerWebProcessState {
 - (NakedPtr<WebKit::WebPageProxy>)_page;
 - (RefPtr<WebKit::WebPageProxy>)_protectedPage;
+#if ENABLE(SCREEN_TIME)
+- (STWebpageController *)screenTimeWebpageController;
```
Need a leading underscore here, because this class is public API and this method is internal.

hortont424 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -379,6 +379,7 @@ OBJC_CLASS WKQLThumbnailLoadOperation;
 OBJC_CLASS WKQuickLookPreviewController;
 OBJC_CLASS WKWebView;
 OBJC_CLASS _WKRemoteObjectRegistry;
+OBJC_CLASS STWebpageController;
```
Sort

hortont424 commented on Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
```diff
@@ -293,6 +297,21 @@ - (BOOL)_hostsLayersInWindowServer;
+
+#if ENABLE(SCREEN_TIME)
+    if (!PAL::isScreenTimeFrameworkAvailable())
+        return;
```
This early return seems a bit dangerous for people adding unrelated code later in this method. Maybe just nest it and invert the if? Or factor it out into another method instead of nesting? I'm not sure.

jesxilin replied:
I think factoring out this logic into another method works better. Early returning in that separate method will not effect code after in `didCommitLoadForMainFrame`! Will change!

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -338,6 +346,60 @@ static uint32_t convertSystemLayoutDirection(NSUserInterfaceLayoutDirection dire
+- (void)_installScreenTimeWebpageController
+{
+    if (!PAL::isScreenTimeFrameworkAvailable())
+        return;
```
minor nit: should probably add an extra newline after this return

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -338,6 +346,60 @@ static uint32_t convertSystemLayoutDirection(NSUserInterfaceLayoutDirection dire
+    if (!PAL::isScreenTimeFrameworkAvailable())
+        return;
+    if (!_page->preferences().screenTimeEnabled())
+        return;
```
ditto

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -338,6 +346,60 @@ static uint32_t convertSystemLayoutDirection(NSUserInterfaceLayoutDirection dire
+
+#if PLATFORM(MAC)
+        NSView *screenTimeView = [_screenTimeWebpageController view];
+        screenTimeView.frame = self.bounds;
```
I don't think this line is needed anymore now that you are using auto layout

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -338,6 +346,60 @@ static uint32_t convertSystemLayoutDirection(NSUserInterfaceLayoutDirection dire
+        [_screenTimeWebpageController addObserver:self forKeyPath:@"URLIsBlocked" options:(NSKeyValueObservingOptionOld | NSKeyValueObservingOptionNew) context:&screenTimeWebpageControllerBlockedKVOContext];
+
+#if PLATFORM(MAC)
+        NSView *screenTimeView = [_screenTimeWebpageController view];
```
RetainPtr screenTimeView = ...

rr-codes commented on Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
```diff
@@ -284,6 +288,29 @@ - (BOOL)_hostsLayersInWindowServer;
+    if (!PAL::isScreenTimeFrameworkAvailable())
+        return;
+
+//    auto webView = webView();
```
should probably remove this comment

jesxilin replied:
omg 💯

rr-codes commented on Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
```diff
@@ -284,6 +288,29 @@ - (BOOL)_hostsLayersInWindowServer;
+
+//    auto webView = webView();
+    NakedPtr<WebKit::WebPageProxy> pageProxy = [webView _page];
+    if (RetainPtr screenTimeWebpageController = [webView _screenTimeWebpageController]) {
```
can probably invert this condition to reduce nesting

rr-codes commented on Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
```diff
@@ -284,6 +288,29 @@ - (BOOL)_hostsLayersInWindowServer;
+            [webView _uninstallScreenTimeWebpageController];
+            return;
+        }
+        screenTimeWebpageController.get().URL = [webView _mainFrameURL];
```
typically we do

```
[screenTimeWebpageController setURL:[webView _mainFrameURL]];
```

rr-codes commented on Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
```diff
@@ -284,6 +288,30 @@ - (BOOL)_hostsLayersInWindowServer;
+    [webView() _installScreenTimeWebpageController];
+}
+
+static void updateScreenTimeWebpageControllerURL(RetainPtr<WKWebView> webView)
```
the parameter should just be WKWebView *webView

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -338,6 +346,62 @@ static uint32_t convertSystemLayoutDirection(NSUserInterfaceLayoutDirection dire
+
+    if (!_screenTimeWebpageController) {
+        _screenTimeWebpageController = adoptNS([PAL::allocSTWebpageControllerInstance() init]);
+        [_screenTimeWebpageController addObserver:self forKeyPath:@"URLIsBlocked" options:(NSKeyValueObservingOptionOld | NSKeyValueObservingOptionNew) context:&screenTimeWebpageControllerBlockedKVOContext];
```
Should this not just be, unless I'm misreading?
```suggestion
        [_screenTimeWebpageController addObserver:self forKeyPath:@"URLIsBlocked" options:(NSKeyValueObservingOptionOld | NSKeyValueObservingOptionNew) context:screenTimeWebpageControllerBlockedKVOContext];
```

jesxilin committed:
[Screen Time Refactoring] macOS support for URL donations
https://bugs.webkit.org/show_bug.cgi?id=283944
rdar://140438800

Reviewed by Aditya Keerthi and Tim Horton.

Add a flag/preference to enable Screen Time donation.
Implemented Screen Time donation for loaded pages.
Installed Screen Time shield view to block web content when needed.
Halts media playback when web content is blocked
and shield is shown (not supported for PIP yet).

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PlatformMac.cmake:
* Source/WebCore/PAL/pal/cocoa/ScreenTimeSoftLink.h: Added.
* Source/WebCore/PAL/pal/cocoa/ScreenTimeSoftLink.mm: Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _installScreenTimeWebpageController]):
(-[WKWebView _uninstallScreenTimeWebpageController]):
(-[WKWebView observeValueForKeyPath:ofObject:change:context:]):
(-[WKWebView dealloc]):
(-[WKWebView _screenTimeWebpageController]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::installScreenTimeWebpageController):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::prepareToLoadWebPage):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::installScreenTimeWebpageController):
(WebKit::updateScreenTimeWebpageControllerURL):
(WebKit::PageClientImpl::didCommitLoadForMainFrame):

Canonical link: https://commits.webkit.org/287638@main
